packages/datadog-instrumentations/src/langchain.js

```diff
@@ -0,0 +1,72 @@
+'use strict'
+
+const { addHook } = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+
+const tracingChannel = require('dc-polyfill').tracingChannel
+
+const invokeTracingChannel = tracingChannel('apm:langchain:invoke')
+
+function wrapLangChainPromise (fn, type, namespace = []) {
+  return function () {
+    if (!invokeTracingChannel.start.hasSubscribers) {
+      return fn.apply(this, arguments)
+    }
+
+    // Runnable interfaces have an `lc_namespace` property
+    const ns = this.lc_namespace || namespace
+    const resource = [...ns, this.constructor.name].join('.')
+
+    const ctx = {
+      args: arguments,
+      instance: this,
+      type,
+      resource
+    }
+
+    return invokeTracingChannel.tracePromise(fn, ctx, this, ...arguments)
+  }
+}
+
+// specifying filePattern: {path}/.*js will match against ESM .js files and CommonJS .cjs files
+// langchain compiles both from their TypeScript source, so we need to match both
+addHook({ name: '@langchain/core', filePattern: 'dist/runnables/base.*js', versions: ['>=0.1'] }, exports => {
+  const RunnableSequence = exports.RunnableSequence
+  shimmer.wrap(RunnableSequence.prototype, 'invoke', invoke => wrapLangChainPromise(invoke, 'chain'))
+  shimmer.wrap(RunnableSequence.prototype, 'batch', batch => wrapLangChainPromise(batch, 'chain'))
+  return exports
+})
+
+addHook({
+  name: '@langchain/core',
+  filePattern: 'dist/language_models/chat_models.*js',
+  versions: ['>=0.1']
+}, exports => {
+  const BaseChatModel = exports.BaseChatModel
+  shimmer.wrap(
+    BaseChatModel.prototype,
+    'generate',
+    generate => wrapLangChainPromise(generate, 'chat_model')
+  )
+  return exports
+})
+
+addHook({ name: '@langchain/core', filePattern: 'dist/language_models/llms.*js', versions: ['>=0.1'] }, exports => {
+  const BaseLLM = exports.BaseLLM
+  shimmer.wrap(BaseLLM.prototype, 'generate', generate => wrapLangChainPromise(generate, 'llm'))
+  return exports
+})
+
+addHook({ name: '@langchain/openai', filePattern: 'dist/embeddings.*js', versions: ['>=0.1'] }, exports => {
+  const OpenAIEmbeddings = exports.OpenAIEmbeddings
+
+  // OpenAI (and Embeddings in general) do not define an lc_namespace
+  const namespace = ['langchain', 'embeddings', 'openai']
+  shimmer.wrap(OpenAIEmbeddings.prototype, 'embedDocuments', embedDocuments =>
+    wrapLangChainPromise(embedDocuments, 'embedding', namespace)
+  )
+  shimmer.wrap(OpenAIEmbeddings.prototype, 'embedQuery', embedQuery =>
+    wrapLangChainPromise(embedQuery, 'embedding', namespace)
+  )
+  return exports
+})
```

Comment:
a small callout here is that we're _only_ patching openai embeddings, and none from `langchain-community`, since it looks like some of the popular ones are deprecated and moved to their own partner library packages, which are harder for us to generically patch. I wasn't able to patch the base `Embeddings` class, as for some reason its compiled definition does not include the `embedQuery` and `embedDocuments` on its prototype. 

should I add embedding tracing for some `langchain-community` packages regardless, or is it OK to start with just openai embeddings for now?

Reply:
IMO fine to start with OpenAI embeddings for now, we can always add more support in the future rather than make this more complicated than necessary atm